core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java

```diff
@@ -0,0 +1,247 @@
+ * {@linkplain ObjectSpooler} allocates two additional objects per parse compared to {#linkplain ObjectParser}. In the worst case it
+ * allocates <code>2 + 2 + 2 * param_count</code> objects per parse. If this overhead is too much for you then feel free to have
+ * ObjectParser parse a secondary object and have that one call the target object's constructor.
+ */
+public final class ObjectSpooler<Value, Context> extends AbstractObjectParser<Value, Context> {
+    private final List<ParseField> constructorArgNames = new ArrayList<>();
+    private final ObjectParser<Spool, Context> objectParser;
+    private final Function<Object[], Value> builder;
+
+    /**
+     * Build the spooler.
+     * 
+     * @param name The name given to the spooler's ObjectParser for error identification. Use what you'd use if the object worked with
+     *        ObjectParser
+     * @param builder A function that builds the object from an array of Objects. Declare this inline with the spooler, casting the elements
+     *        of the array to the arguments so they work with your favorite constructor. The objects in the array will be in the same order
+     *        that you declared the {{@link #constructorArg(String)}s and none will be null. If any of the constructor arguments aren't
+     *        defined in the XContent then parsing will throw an error. We use an array here rather than a Map to save on allocations.
+     */
+    public ObjectSpooler(String name, Function<Object[], Value> builder) {
+        objectParser = new ObjectParser<>(name);
+        this.builder = builder;
+    }
+
+    /**
+     * Call this to do the actual parsing. This implements {@link BiFunction} for conveniently integrating with ObjectParser.
+     */
+    @Override
+    public Value apply(XContentParser parser, Context context) {
+        try {
+            return objectParser.parse(parser, new Spool(parser), context).finish();
+        } catch (IOException e) {
+            throw new ParsingException(parser.getTokenLocation(), "[" + objectParser.getName()  + "] failed to parse object", e);
+        }
+    }
+
+    /**
+     * Pass the {@linkplain BiConsumer} this returns to {@link #parser()}'s declare methods to declare a constructor argument. See this
+     * class's javadoc for an example. The order in which these are declared matters: it is the order that they come in the array passed to
+     * {@link #builder} and the order that missing arguments are reported to the user if any are missing. When all of these parameters are
+     * parsed from the {@linkplain XContentParser} the target object is immediately built.
+     */
+    public <FieldT> BiConsumer<Value, FieldT> constructorArg() {
+        return new ConstructorArg<>();
+    }
+
+    @Override
+    public <T> void declareField(BiConsumer<Value, T> consumer, ContextParser<Context, T> parser, ParseField parseField, ValueType type) {
```

Comment:
The second option here would actually mean that we wouldn't need the `ConstructorArg` class or the `this.constructorArg()` method since its only a marker anyway and the `declareConstructorField()` method would then act as the same marker